test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -35,7 +36,203 @@ import (
 	"github.com/jetstack/cert-manager/test/integration/framework"
 )
 
-func TestCtlCreateCR(t *testing.T) {
+type CreateCRTest struct {
+	inputFile      string
+	inputArgs      []string
+	inputNamespace string
+	keyFilename    string
+	certFilename   string
+	fetchCert      bool
+	timeout        time.Duration
+	crStatus       cmapiv1alpha2.CertificateRequestStatus
+
+	expRunErr          bool
+	expErrMsg          string
+	expNamespace       string
+	expName            string
+	expKeyFilename     string
+	expCertFilename    string
+	expCertFileContent []byte
+}
+
+// This test suite is split into multiple functions to make it easier to test the `create cr` command
+// more granularly.
+// For example if the command is expected to fail after the private key has been written to file,
+// but before the CR is successfully created, then we still want to check the validity of the private key.
+
+// TestCtlCreateCRBehaviourBeforeAnythingIsCreated tests the behaviour in the case where the command fails
+// before anything is created, i.e. before private key is successfully written to file.
+func TestCtlCreateCRBehaviourBeforeAnythingIsCreated(t *testing.T) {
+	config, stopFn := framework.RunControlPlane(t)
+	defer stopFn()
+
+	// Build clients
+	_, _, cmCl, _ := framework.NewClients(t, config)
+
+	testdataPath := getTestDataDir(t)
+
+	const (
+		cr3Name = "testcr-3"
+		cr4Name = "testcr-4"
+		ns1     = "testns-1"
+		ns2     = "testns-2"
+	)
+
+	tests := map[string]CreateCRTest{
+		"conflicting namespaces defined in flag and file": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr3Name},
+			inputNamespace: ns2,
+			keyFilename:    "",
+			expRunErr:      true,
+			expErrMsg:      "the namespace from the provided object \"testns-1\" does not match the namespace \"testns-2\". You must pass '--namespace=testns-1' to perform this operation.",
+		},
+		"file passed in defines resource other than certificate": {
+			inputFile:      testdataPath + "create_cr_issuer.yaml",
+			inputArgs:      []string{cr4Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expRunErr:      true,
+			expErrMsg:      "decoded object is not a v1alpha2 Certificate",
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			cleanUpFunc := setupPathForTest(t)
+			defer cleanUpFunc()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
```

Comment:
Same comment as before?

Reply:
Oh right, you wanted to change both the unit test and the integration tests. Same as before, I don't see how one way is better than the other.